src/MultiFormats/MultiAddress.cs

```diff
@@ -0,0 +1,565 @@
+        public MultiAddress(byte[] buffer)
+            : this()
+        {
+            if (buffer == null || buffer.Length == 0)
+            {
+                return;
+            }
+
+            Read(new MemoryStream(buffer, false));
+        }
+
+        /// <summary>
+        ///   Gets the peer ID of the multiaddress.
+        /// </summary>
+        /// <value>
+        ///   The <see>
+        ///       <cref>Lib.P2P.Peer.Id</cref>
+        ///   </see>
+        ///   as a <see cref="MultiHash"/>.
+        /// </value>
+        /// <exception cref="Exception">
+        ///   When the last <see cref="Protocols">protocol</see>
+        ///   is not "ipfs" nor "p2p".
+        /// </exception>
+        /// <remarks>
+        ///   The peer ID is contained in the last protocol that
+        ///   is "ipfs" or "p2p".  For example, <c>/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC</c>.
+        /// </remarks>
+        public MultiHash PeerId
+        {
+            get
+            {
+                var protocol = Protocols
+                   .LastOrDefault(p => p.Name == "ipfs" || p.Name == "p2p");
+                if (protocol == null)
+                {
+                    throw new Exception($"'{this}' is missing the peer ID. Add the 'ipfs' or 'p2p' protocol.");
+                }
+
+                return protocol.Value;
+            }
+        }
+
+        /// <summary>
+        ///   Determines if the peer ID is present.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> if the peer ID present; otherwise, <b>false</b>.
+        /// </value>
+        /// <remarks>
+        ///   The peer ID is contained in the last protocol that
+        ///   is "ipfs" or "p2p".  For example, <c>/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC</c>.
+        /// </remarks>
+        public bool HasPeerId { get { return Protocols.Any(p => p.Name == "ipfs" || p.Name == "p2p"); } }
+
+        /// <summary>
+        ///   Gets a multiaddress that ends with the peer ID.
+        /// </summary>
+        /// <param name="peerId">
+        ///   The peer ID to end the multiaddress with.
+        /// </param>
+        /// <returns>
+        ///   Either the <c>this</c> multiadddress when it contains the
+        ///   <paramref name="peerId"/> or a new <see cref="MultiAddress"/>
+        ///   ending the <paramref name="peerId"/>.
+        /// </returns>
+        /// <exception cref="Exception">
+        ///   When the mulltiaddress has the wrong peer ID.
+        /// </exception>
+        public MultiAddress WithPeerId(MultiHash peerId)
+        {
+            if (!HasPeerId)
+            {
+                return new MultiAddress(ToString() + $"/p2p/{peerId}");
+            }
+            
+            var id = PeerId;
+            if (id != peerId)
+            {
+                throw new Exception($"Expected a multiaddress with peer ID of '{peerId}', not '{id}'.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)